In [1]:
#referred to lesson 9.05
#https://stackoverflow.com/questions/61757170/python-unstacked-dataframe-is-too-big-causing-int32-overflow
#https://stackoverflow.com/questions/28651079/pandas-unstack-problems-valueerror-index-contains-duplicate-entries-cannot-re#:~:text=The%20reason%20why%20you%20get%20ValueError%3A%20Index%20contains,%22%20date%20%22%20combinations%20are%20no%20longer%20unique.
#https://stackoverflow.com/questions/43945653/python-pandas-return-dataframe-where-value-count-is-above-a-set-number
#https://stackoverflow.com/questions/54822879/how-can-i-combine-multiple-sparse-and-dense-matrices-together
#got help from Dan Wilhelm (GA) on creating sparse matrix
#https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html
import pandas as pd
import numpy as np
from scipy import sparse
from sklearn.metrics.pairwise import pairwise_distances, cosine_distances, cosine_similarity

In [2]:
df=pd.read_csv('./data/video_games.csv')

In [3]:
df.isnull().sum()

customer_id           0
review_id             0
product_id            0
product_parent        0
product_title         0
star_rating           0
helpful_votes         0
total_votes           0
verified_purchase     0
review_date           0
full_review          47
dtype: int64

In [4]:
df.dropna(inplace=True)

In [5]:
df['review_date'] = pd.to_datetime(df['review_date'])

In [6]:
df.head()

customer_id       review_id  product_id  product_parent  \
0     12039526   RTIS3L2M1F5SM  B001CXYMFS       737716809   
1      2331478  R3BH071QLH8QMC  B0029CSOD2        98937668   
2     52495923  R127K9NTSXA2YH  B00GOOSV98        23143350   
3     14533949  R32ZWUXDJPW27Q  B00Y074JOM       821342511   
4     17521011  R2F0POU5K6F73F  B008XHCLFO        24234603   

                                       product_title  star_rating  \
0         Thrustmaster T-Flight Hotas X Flight Stick            5   
1  Hidden Mysteries: Titanic Secrets of the Fatef...            1   
2  GelTabz Performance Thumb Grips - PlayStation ...            3   
3  Zero Suit Samus amiibo - Japan Import (Super S...            4   
4                         Protection for your 3DS XL            5   

   helpful_votes  total_votes  verified_purchase review_date  \
0              0            0                  1  2015-08-31   
1              0            1                  1  2015-08-31   
2              0            0                  1  2015-08-31   
3              0            0                  1  2015-08-31   
4              0            0                  1  2015-08-31   

                                         full_review  
0  amazing joystick I especially love twist Used ...  
1              One Star poor quality work advertised  
2  good could bettee nice tend slip away stick in...  
3  Great flawed Great amiibo great collecting Qua...  
4  A Must I 2012 2013 XL durable comfortable real...

In [7]:
df.columns

Index(['customer_id', 'review_id', 'product_id', 'product_parent',
       'product_title', 'star_rating', 'helpful_votes', 'total_votes',
       'verified_purchase', 'review_date', 'full_review'],
      dtype='object')

I need to see if any customers wrote multiple reviews for the same product. That would throw an error in the pivot table.

In [8]:
df.groupby('customer_id')['product_id'].value_counts().sort_values(ascending=False)

customer_id  product_id
38142327     B00005NH6B    13
43411792     B00000JRSB    13
50763007     B0000AHOOJ    10
25214010     B0000C4M22    10
30995260     B00005BW6Z    10
                           ..
38010719     B000035Y6N     1
38010773     B001C91H4G     1
             B001DL8PES     1
38010873     B0083RDT8C     1
10018        B006VE40JQ     1
Name: product_id, Length: 1641949, dtype: int64

Hm... for example, customer 38142327 wrote 13 reviews for product B00005NH6B... what do those reviews look like?

In [12]:
df.loc[(df['product_id']=='B00005NH6B') & (df['customer_id']==38142327)]

customer_id       review_id  product_id  product_parent  \
1571527     38142327  R3NY8KHMR5X2KP  B00005NH6B       199086987   
1572890     38142327  R2U1NNLJS8OHHQ  B00005NH6B       199086987   
1572901     38142327  R122RU8UP19R4T  B00005NH6B       199086987   
1572912     38142327  R22Z951P3JMRBQ  B00005NH6B       199086987   
1572927     38142327  R2RCRESG64DH7D  B00005NH6B       199086987   
1572929     38142327   RHMREFYJF0IWP  B00005NH6B       199086987   
1572933     38142327  R1TZYG1TID6GE6  B00005NH6B       199086987   
1572935     38142327  R1QDQSKQ7XJPLI  B00005NH6B       199086987   
1572936     38142327  R2RI3Y4DAWBISL  B00005NH6B       199086987   
1572979     38142327   RVES7ZUE7F43B  B00005NH6B       199086987   
1572980     38142327  R2AN652064BA96  B00005NH6B       199086987   
1572983     38142327  R1MMLLRVDR3XS5  B00005NH6B       199086987   
1573111     38142327  R32VTLFJ3E67H6  B00005NH6B       199086987   

            product_title  star_rating  helpful_votes  total_votes  \
1571527  Batman Vengeance            1              1            2   
1572890  Batman Vengeance            5              0            1   
1572901  Batman Vengeance            5              0            0   
1572912  Batman Vengeance            5              0            0   
1572927  Batman Vengeance            5              1            1   
1572929  Batman Vengeance            4              0            2   
1572933  Batman Vengeance            4              1            2   
1572935  Batman Vengeance            5              0            1   
1572936  Batman Vengeance            5              1            2   
1572979  Batman Vengeance            4              0            0   
1572980  Batman Vengeance            5              1            1   
1572983  Batman Vengeance            5              0            0   
1573111  Batman Vengeance            5              4            4   

         verified_purchase review_date  \
1571527                  0  2002-06-29   
1572890                  0  2002-06-17   
1572901                  0  2002-06-17   
1572912                  0  2002-06-17   
1572927                  0  2002-06-17   
1572929                  0  2002-06-17   
1572933                  0  2002-06-17   
1572935                  0  2002-06-17   
1572936                  0  2002-06-17   
1572979                  0  2002-06-17   
1572980                  0  2002-06-17   
1572983                  0  2002-06-17   
1573111                  0  2002-06-16   

                                               full_review  
1571527  Pitiful I huge Batman fan Anything Batman bet ...  
1572890  Batman come alooong way From 1930 new milleniu...  
1572901  FUN FOR FAMILY If like playing game family spe...  
1572912  Best game hit store There isnt superhero like ...  
1572927  Fairly good game kid 12 First I would like say...  
1572929  WHATCHA LOOKIN AT WAAAAAAAAAAAAAAAAAAAAAAAAAAA...  
1572933  ha spider man beat I really dont feel like wri...  
1572935  You sorry miss UBI Soft best Good work Warner ...  
1572936  Best yet If PS2 nice fun game get It pass time...  
1572979  One DC Comic greatest superheroes finally go P...  
1572980  MAKES PS2 THE BEST GAME CONSOLE Cool game Cool...  
1572983  ONE NOT TO MISS ABSOLUTELY NOTHING BUT FUN PER...  
1573111  Best Super Hero Game PS2 console I mean game a...

So this person made 13 reviews for the same item, rating 4 or 5 stars EXCEPT for the most recent one (6/29) which was 1 star. Perhaps the ending was disappointing?

In [13]:
review_bools = df.groupby('customer_id')['product_id'].value_counts()>1
#T/F if more than 1 review for same product

In [14]:
review_bools[review_bools==True].count()

5309

Okay, based on this there are 5,309 customers with at least 2 reviews for the same product.  

I could try to get the most recent review for each customer based on date. It's likely reasonable to assume that their latest review is probably the one that best reflects their final sentiment about the product. Unfortunately, because I don't have precise timestamps, if they wrote two reviews on the same day, I can't know which was their last word on the product. Other options would be to average all of their reviews' star values or to simply drop them, as they're a fairly small proportion of the overall set.

In [15]:
xtra_rev_cs = []
for key, value in dict(review_bools[review_bools==True]).items():
    xtra_rev_cs.append({key[0]:key[1]}) #return customer id and product id ONLY
#list of customer id numbers for those with more than 1 review for same item
xtra_rev_cs[:5]  #preview

[{198273: 'B00DNGQQUQ'},
 {224068: 'B0049LAVB4'},
 {815919: 'B0000TSR4C'},
 {884129: 'B000WMEEB2'},
 {884129: 'B00ERDGMT4'}]

In [16]:
df[(df['customer_id'] == 38142327) & 
   (df['product_id']=='B00005NH6B')].sort_values(
    by='review_date', ascending=False).index[1:] #test

Int64Index([1572890, 1572901, 1572912, 1572927, 1572929, 1572933, 1572935,
            1572936, 1572979, 1572980, 1572983, 1573111],
           dtype='int64')

Okay - with the above code in conjunction with the xtra_rev_cs dictionary, I can filter the dataframe by customer_id and product_id to isolate the rows that pertain to a single product reviewed mutliple times by the same user, and then order it by date where the most recent date will be on top. While this does open the possibility that two reviews with the same date may be conflated as far as which ACTUALLY came first, the examples I've previewed in this situation are typically duplicates (i.e. someone submitted the same review twice). In that circumstance, either review is equally valid to keep.  

The next step is to loop through my xtra_rev_cs dictionary and create a list of all the indexes EXCEPT the most recent one. I'll drop those others, and be left with one review per person per item.

In [17]:
rev_indexes_to_drop = []
for pair in xtra_rev_cs:
    for key, value in pair.items():
        rev_indexes_to_drop.append(  #add index numbers to the empty list
        df[(df['customer_id'] == key) &    #where customer id is the key from xtra_rev_cs...
       (df['product_id']==value)].sort_values( #and product id is the value from xtra_rev_cs
        by='review_date', ascending=False).index[1:] #starting with the SECOND index number
        )
rev_indexes_to_drop[:5] #preview

[Int64Index([26651], dtype='int64'),
 Int64Index([70501], dtype='int64'),
 Int64Index([90766], dtype='int64'),
 Int64Index([422385], dtype='int64'),
 Int64Index([421883], dtype='int64')]

In [18]:
rev_indexes_to_drop[0][0] #test

26651

In [19]:
ritd_2 = []
for n in rev_indexes_to_drop:
    for k in n:
        ritd_2.append(k)
ritd_2[:5] #preview

[26651, 70501, 90766, 422385, 421883]

In [20]:
len(ritd_2) #6140 rows of duplicate reviews to be dropped

6140

In [21]:
df.drop(index=ritd_2, inplace=True) #drop all index numbers in list ritd_2
df.groupby('customer_id')['product_id'].value_counts().sort_values(ascending=False) #check if gone

customer_id  product_id
53096565     B00006599W    1
17664004     B000U5RRX8    1
17663554     B000SH3XGS    1
17663609     B002TK1PX0    1
17663621     B00E20STAW    1
                          ..
38073024     B0088MVOES    1
38073066     B006VB2UNM    1
38073092     B007P6Y684    1
             B00A9ZHWH0    1
10018        B006VE40JQ    1
Name: product_id, Length: 1641949, dtype: int64

In [22]:
df2 = df[['customer_id', 'product_id', 'product_title', 'star_rating']].copy()

In [30]:
#pivot = pd.pivot_table(df2, index='product_title', columns='customer_id', values='star_rating')

My first attempt at making a pivot table from this data failed because the created dataframe is too big. I'll have to do it in chunks.

In [64]:
#got help from Dan Wilhelm on creating sparse matrix

In [31]:
df2.shape #rows

(1641949, 4)

In [32]:
df2['customer_id'].nunique() #unique customers

979917

In [33]:
df2['product_id'].nunique() #unique products

20952

In [24]:
len(df2)/37

44377.0

In [25]:
df2[:10_000] #partial dataframe

customer_id  product_id  \
0         12039526  B001CXYMFS   
1          2331478  B0029CSOD2   
2         52495923  B00GOOSV98   
3         14533949  B00Y074JOM   
4         17521011  B008XHCLFO   
...            ...         ...   
9996       2597447  B00005LVWL   
9997      34800440  B00004WFVZ   
9998      13984861  B00VETEZ34   
9999      21531468  B00KAED850   
10000     34899134  B00A7Q9ZQM   

                                           product_title  star_rating  
0             Thrustmaster T-Flight Hotas X Flight Stick            5  
1      Hidden Mysteries: Titanic Secrets of the Fatef...            1  
2      GelTabz Performance Thumb Grips - PlayStation ...            3  
3      Zero Suit Samus amiibo - Japan Import (Super S...            4  
4                             Protection for your 3DS XL            5  
...                                                  ...          ...  
9996        Nancy Drew: Treasure In The Royal Tower - PC            5  
9997   Microsoft Combat Flight Simulator 2:  Pacific ...            5  
9998   PlayStation 4 500GB Console - Batman Arkham Kn...            5  
9999                                           Far Cry 4            5  
10000                            Tiger Woods PGA TOUR 14            4  

[10000 rows x 4 columns]

In [39]:
df2.dtypes

customer_id       int64
product_id       object
product_title    object
star_rating       int64
dtype: object

My matrix will need the title input to be a numerical value (int); I will create a list of unique products and match them with a number 1 - (# of products) in a dictionary, and then add a new column with that number to my dataframe.

In [76]:
unique_prods = list(set(df2['product_title']))
prod_index = {p:i for i,p in enumerate(unique_prods)}

In [78]:
df2['prod_numerical'] = df2['product_title'].apply(lambda x: prod_index[x])

In [79]:
df2.head()

customer_id  product_id                                      product_title  \
0     12039526  B001CXYMFS         Thrustmaster T-Flight Hotas X Flight Stick   
1      2331478  B0029CSOD2  Hidden Mysteries: Titanic Secrets of the Fatef...   
2     52495923  B00GOOSV98  GelTabz Performance Thumb Grips - PlayStation ...   
3     14533949  B00Y074JOM  Zero Suit Samus amiibo - Japan Import (Super S...   
4     17521011  B008XHCLFO                         Protection for your 3DS XL   

   star_rating  prod_numerical  cs_numerical  
0            5            8234        727492  
1            1           12127        113959  
2            3            8137         32535  
3            4            1943        912813  
4            5            4512        358299

I should try to save memory in other ways too. I can convert my start rating to an 8bit format since it will only ever be numbers 1-5. I can also convert my unique customer id numbers (already integers) to smaller numbers so that my matrix doesn't make empty rows in between them.

In [80]:
df2['star_rating'] = df2['star_rating'].astype(np.int8) #convert to take less memory

In [81]:
unique_cs = list(set(df2['customer_id']))
cs_index = {p:i for i,p in enumerate(unique_cs)}
df2['cs_numerical'] = df2['customer_id'].apply(lambda x: cs_index[x])

In [82]:
df2.head()

customer_id  product_id                                      product_title  \
0     12039526  B001CXYMFS         Thrustmaster T-Flight Hotas X Flight Stick   
1      2331478  B0029CSOD2  Hidden Mysteries: Titanic Secrets of the Fatef...   
2     52495923  B00GOOSV98  GelTabz Performance Thumb Grips - PlayStation ...   
3     14533949  B00Y074JOM  Zero Suit Samus amiibo - Japan Import (Super S...   
4     17521011  B008XHCLFO                         Protection for your 3DS XL   

   star_rating  prod_numerical  cs_numerical  
0            5            8234        727492  
1            1           12127        113959  
2            3            8137         32535  
3            4            1943        912813  
4            5            4512        358299

Okay - that's probably about as minimal as we can get our inputs; let's make this sparse matrix!

In [83]:
sparse_reviews = sparse.csr_matrix((df2.star_rating, (df2.prod_numerical, df2.cs_numerical)), dtype=np.int8)

In [84]:
sparse_reviews.shape

(15938, 979917)

In [87]:
print(sparse_reviews[1]) #preview

  (0, 18975)	5
  (0, 58323)	5
  (0, 98068)	5
  (0, 145194)	5
  (0, 162282)	5
  (0, 198137)	3
  (0, 263742)	5
  (0, 286623)	5
  (0, 315515)	5
  (0, 397925)	5
  (0, 401274)	4
  (0, 428441)	5
  (0, 520336)	5
  (0, 569002)	5
  (0, 571384)	5
  (0, 698680)	5
  (0, 752957)	5
  (0, 791097)	5
  (0, 803120)	5
  (0, 815823)	5
  (0, 827741)	5
  (0, 851608)	3
  (0, 866826)	5
  (0, 926253)	1
  (0, 928136)	5
  (0, 931576)	5
  (0, 955110)	5
  (0, 963893)	3
  (0, 970708)	5
  (0, 974420)	5


In [88]:
df2[df2['prod_numerical']==1] #test - cs numbers above should match cs_numerical column below

customer_id  product_id  \
6686        41930983  B002GCJOC0   
11451        8686286  B002GCJOC0   
29126        1214997  B002GCJOC0   
33619       26773768  B002GCJOC0   
34662        2042898  B002GCJOC0   
44088       36548550  B002GCJOC0   
44868       46726639  B002GCJOC0   
83340       23270727  B002GCJOC0   
94019       28100332  B002GCJOC0   
94853        2218072  B002GCJOC0   
120464       1741699  B002GCJOC0   
129285       1209522  B002GCJOC0   
131922      39515894  B002GCJOC0   
137983       1818079  B002GCJOC0   
158089        332268  B002GCJOC0   
391210       1980578  B002GCJOC0   
433802      28756851  B002GCJOC0   
434868       4159550  B002GCJOC0   
435319      28092822  B002GCJOC0   
472897      29114056  B002GCJOC0   
495160      43657579  B002GCJOC0   
507890      25060969  B002GCJOC0   
512140      35243730  B002GCJOC0   
687125       8276240  B002GCJOC0   
757149      15222096  B002GCJOC0   
793199      15085560  B002GCJOC0   
932376      15329907  B002GCJOC0   
970430      12562740  B002GCJOC0   
1048506     43044346  B002GCJOC0   
1253272     41982487  B002GCJOC0   

                                        product_title  star_rating  \
6686     AV Audio / Video Cable for Nintendo Gamecube            5   
11451    AV Audio / Video Cable for Nintendo Gamecube            5   
29126    AV Audio / Video Cable for Nintendo Gamecube            5   
33619    AV Audio / Video Cable for Nintendo Gamecube            5   
34662    AV Audio / Video Cable for Nintendo Gamecube            5   
44088    AV Audio / Video Cable for Nintendo Gamecube            5   
44868    AV Audio / Video Cable for Nintendo Gamecube            5   
83340    AV Audio / Video Cable for Nintendo Gamecube            5   
94019    AV Audio / Video Cable for Nintendo Gamecube            4   
94853    AV Audio / Video Cable for Nintendo Gamecube            5   
120464   AV Audio / Video Cable for Nintendo Gamecube            5   
129285   AV Audio / Video Cable for Nintendo Gamecube            5   
131922   AV Audio / Video Cable for Nintendo Gamecube            5   
137983   AV Audio / Video Cable for Nintendo Gamecube            3   
158089   AV Audio / Video Cable for Nintendo Gamecube            5   
391210   AV Audio / Video Cable for Nintendo Gamecube            1   
433802   AV Audio / Video Cable for Nintendo Gamecube            5   
434868   AV Audio / Video Cable for Nintendo Gamecube            3   
435319   AV Audio / Video Cable for Nintendo Gamecube            5   
472897   AV Audio / Video Cable for Nintendo Gamecube            5   
495160   AV Audio / Video Cable for Nintendo Gamecube            5   
507890   AV Audio / Video Cable for Nintendo Gamecube            5   
512140   AV Audio / Video Cable for Nintendo Gamecube            5   
687125   AV Audio / Video Cable for Nintendo Gamecube            5   
757149   AV Audio / Video Cable for Nintendo Gamecube            5   
793199   AV Audio / Video Cable for Nintendo Gamecube            3   
932376   AV Audio / Video Cable for Nintendo Gamecube            5   
970430   AV Audio / Video Cable for Nintendo Gamecube            5   
1048506  AV Audio / Video Cable for Nintendo Gamecube            5   
1253272  AV Audio / Video Cable for Nintendo Gamecube            5   

         prod_numerical  cs_numerical  
6686                  1        974420  
11451                 1        145194  
29126                 1        571384  
33619                 1        752957  
34662                 1        955110  
44088                 1        428441  
44868                 1        286623  
83340                 1         98068  
94019                 1        401274  
94853                 1         58323  
120464                1        815823  
129285                1        569002  
131922                1        827741  
137983                1        851608  
158089                1        162282  
391210                1        926253  
433802                1        698680  
4

The matrix looks good - now it's time to make the recommender model. In order to do this we will look into the cosine similarities between items to see which ones are on similar vectors i.e. liked in equal measure to other items by the same users.

In [89]:
from sklearn.metrics.pairwise import pairwise_distances, cosine_distances, cosine_similarity

In [90]:
dists = pairwise_distances(sparse_reviews, metric='cosine')

In [91]:
dists #preview

array([[0.        , 1.        , 0.98175221, ..., 1.        , 1.        ,
        1.        ],
       [1.        , 0.        , 1.        , ..., 1.        , 1.        ,
        1.        ],
       [0.98175221, 1.        , 0.        , ..., 1.        , 1.        ,
        0.9933939 ],
       ...,
       [1.        , 1.        , 1.        , ..., 0.        , 1.        ,
        1.        ],
       [1.        , 1.        , 1.        , ..., 1.        , 0.        ,
        1.        ],
       [1.        , 1.        , 0.9933939 , ..., 1.        , 1.        ,
        0.        ]])

In [92]:
recommender_df = pd.DataFrame(dists,
                             index=unique_prods,
                             columns=unique_prods)

recommender_df.head()

Ape Escape  \
Ape Escape                                      0.000000   
AV Audio / Video Cable for Nintendo Gamecube    1.000000   
Fantastic Four                                  0.981752   
Serious Sam 2 - Xbox                            1.000000   
Cabela's Adventure Camp - Nintendo Wii          1.000000   

                                              AV Audio / Video Cable for Nintendo Gamecube  \
Ape Escape                                                                             1.0   
AV Audio / Video Cable for Nintendo Gamecube                                           0.0   
Fantastic Four                                                                         1.0   
Serious Sam 2 - Xbox                                                                   1.0   
Cabela's Adventure Camp - Nintendo Wii                                                 1.0   

                                              Fantastic Four  \
Ape Escape                                          0.981752   
AV Audio / Video Cable for Nintendo Gamecube        1.000000   
Fantastic Four                                      0.000000   
Serious Sam 2 - Xbox                                1.000000   
Cabela's Adventure Camp - Nintendo Wii              1.000000   

                                              Serious Sam 2 - Xbox  \
Ape Escape                                                     1.0   
AV Audio / Video Cable for Nintendo Gamecube                   1.0   
Fantastic Four                                                 1.0   
Serious Sam 2 - Xbox                                           0.0   
Cabela's Adventure Camp - Nintendo Wii                         1.0   

                                              Cabela's Adventure Camp - Nintendo Wii  \
Ape Escape                                                                       1.0   
AV Audio / Video Cable for Nintendo Gamecube                                     1.0   
Fantastic Four                                                                   1.0   
Serious Sam 2 - Xbox                                                             1.0   
Cabela's Adventure Camp - Nintendo Wii                                           0.0   

                                              Bust-A-Move Bash!  \
Ape Escape                                                  1.0   
AV Audio / Video Cable for Nintendo Gamecube                1.0   
Fantastic Four                                              1.0   
Serious Sam 2 - Xbox                                        1.0   
Cabela's Adventure Camp - Nintendo Wii                      1.0   

                                              Prison Break: The Conspiracy  \
Ape Escape                                                             1.0   
AV Audio / Video Cable for Nintendo Gamecube                           1.0   
Fantastic Four                                                         1.0   
Serious Sam 2 - Xbox                                                   1.0   
Cabela's Adventure Camp - Nintendo Wii                                 1.0   

                                              Star Wars Battlefront: Renegade Squadron - Sony PSP  \
Ape Escape                                                                                  1.0     
AV Audio / Video Cable for Nintendo Gamecube                                                1.0     
Fantastic Four                                                                              1.0     
Serious Sam 2 - Xbox                                                                        1.0     
Cabela's Adventure Camp - Nintendo Wii                                                      1.0     

                                              Playstation 2 Magna Force 2.4Ghz RF Wireless Controller Pink  \
Ape Escape                                                                                  1.0              
AV Audio / Video Cable for Nintendo Gamecube                                          

Alright! There's our recommender dataframe - the numbers in the cells show how related the titles are to each other - 0 means perfect match as a measure of distance (that's why 0's are where the item is matched with itself) and then 1 is the furthest distance away.

In [ ]:
q = 'Mass Effect'
titles = list(df2[df2['product_title'].str.contains(q)]['product_title'])
for title in titles:
    

In [113]:
q = 'Mass Effect'
titles = list(df2[df2['product_title'].str.contains(q)]['product_title'])
#for title in titles:
    #print(title)
    #print('Average rating:', df2.['title'.mean())
    #print('Number of ratings:', df2.loc[title,:].count())
    #print()
print('10 closest items:', recommender_df.loc[titles[0],:].sort_values()[0:10])
    #print('\n***************************************\n')

10 closest items: Mass Effect Trilogy                         0.000000
PS3 Killzone Trilogy Collection - 2 Disc    0.974535
Dragon Age Origins: Ultimate Edition        0.977305
inFAMOUS Collection - Playstation 3         0.979790
Tomb Raider                                 0.980698
Castlevania: Lords of Shadow 2              0.981201
Dragon Age 2                                0.981485
Assassin's Creed Ezio Trilogy Edition       0.981518
Inversion                                   0.981521
Wii Rayman Raving Rabbids Dual Charger      0.981600
Name: Mass Effect Trilogy, dtype: float64
